### PR TITLE
Use cached dnnl::stream in GpuStreamManager

### PR DESCRIPTION
**Pull request by guangyey**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #139176

# Motivation
The code changes in `GpuStreamManager` class intend to help manage `dnnl::stream` efficiently.

# Addtional Context
Use the following code to simply benchmark.
```python
import torch
import time

device = torch.device("xpu")

M, N, K = 64, 64, 64  # You can change these dimensions as needed
torch.manual_seed(0)

A = torch.randn(M, K, device=device)
B = torch.randn(K, N, device=device)

# Warm-up
for _ in range(10):
    torch.matmul(A, B)

s1 = torch.xpu.Stream()
s2 = torch.xpu.Stream()

# Measure the time for the GEMM operation
start_time = time.time()
with torch.xpu.stream(s1):
    for _ in range(50000):
        C = torch.matmul(A, B)

with torch.xpu.stream(s2):
    for _ in range(50000):
        D = torch.matmul(A, B)

torch.xpu.synchronize()
end_time = time.time()

# Calculate elapsed time
elapsed_time = end_time - start_time

# Print the results
print(f"Time taken for GEMM operation: {elapsed_time:.6f} seconds")
```
Compared with the old implementation elapses 2.077069s, the new implementation consumes 2.023017s, which means ~2% performance improvement.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @gujinghui @EikanWang @fengyuan14

### CODE REVIEW

File: aten/src/ATen/native/mkldnn/xpu/detail/oneDNNContext.h [hunk 1]
```diff
@@ -55,21 +55,37 @@ struct TORCH_XPU_API GpuStreamManager {
   static GpuStreamManager& Instance(); // Singleton
 
   dnnl::stream get_stream() {
-    c10::DeviceIndex device_index = c10::xpu::current_device();
-    TORCH_INTERNAL_ASSERT(device_index < c10::xpu::device_count());
-    return dnnl::sycl_interop::make_stream(
```
Comment (gujinghui) [inline aten/src/ATen/native/mkldnn/xpu/detail/oneDNNContext.h:60]: Why did not upstream stream pool before? Confused...

Comment (guangyey) [reply aten/src/ATen/native/mkldnn/xpu/detail/oneDNNContext.h:60]: As discussed offline, we use an elegant method to implement a stream pool to ensure compatibility with future feature `external stream`.

**Review by gujinghui — APPROVED** [review]

**Review by jgong5 — APPROVED** [review]

**Pull request closed without merging**